src/v/cloud_storage/topic_manifest.cc

```diff
@@ -22,12 +23,175 @@
 
 #include <boost/lexical_cast.hpp>
 #include <fmt/ostream.h>
+#include <rapidjson/encodings.h>
+#include <rapidjson/error/en.h>
 #include <rapidjson/istreamwrapper.h>
 #include <rapidjson/ostreamwrapper.h>
+#include <rapidjson/reader.h>
 
 #include <chrono>
+#include <stdexcept>
+#include <string_view>
 
 namespace cloud_storage {
+
+struct topic_manifest_handler
+  : public rapidjson::
+      BaseReaderHandler<rapidjson::UTF8<>, topic_manifest_handler> {
+    topic_manifest_handler()
+      : _state(expectManifestStart) {}
+
+    bool StartObject() {
+        vlog(cst_log.trace, "StartObject(), state_: {}", _state);
+        switch (_state) {
+        case expectManifestStart:
+            _state = expectKey;
+            return true;
+        default:
+            return false;
+        }
+    }
+
+    bool Key(const char* str, rapidjson::SizeType length, bool copy) {
+        vlog(
+          cst_log.trace,
+          "Key({}, {}, {}), state_: {}",
+          str,
+          length,
+          copy,
+          _state);
+        switch (_state) {
+        case expectKey:
+        case expectKeyOrObjectEnd:
+            _key = std::string(str, length);
```

Comment:
we can use `ss::sstring`s for keys and specify the template small string size to be 31, so that even the "cleanup_policy_bitflags" key string doesn't allocate.

Reply:
If `copy` is `true`, then some kind of copy is required. If parsing is done in-place, `copy` will be `false`, and a `string_view` would suffice. Parsing in-place may be tricky, especially under streaming scenarios.

The way that I've done it in the past is to have a state for each known key.

Reply:
Changed to use `ss::basic_sstring<char, uint32_t, 31>`